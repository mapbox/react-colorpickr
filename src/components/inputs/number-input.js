'use strict';

import React from 'react';
import PropTypes from 'prop-types';

class NumberInput extends React.Component {
  static propTypes = {
    label: PropTypes.string.isRequired,
    value: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired,
    min: PropTypes.number.isRequired,
    max: PropTypes.number.isRequired
  };

  render() {
    const { label, value, onChange, min, max } = this.props;
    return (
      <div className='flex-child flex-child--grow relative'>
        <label className='absolute top left pl6 py3 color-gray-light txt-bold'>{label}</label>
        <input className='w-full pl18 input input--s bg-white' value={value} onChange={onChange} type="number" min={min} max={max} step={1} />
      </div>
    );
  }
}

export default NumberInput;
