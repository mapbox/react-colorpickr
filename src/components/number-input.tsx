import React, { useState, useCallback, useEffect } from 'react';
import themeable from 'react-themeable';
import { autokey } from '../autokey.ts';

interface Props {
  id: string;
  value: number;
  theme: { [id: string]: string };
  onChange: (id: string, value: number) => void;
  min: number;
  max: number;
  readOnly?: boolean;
}

function NumberInput({
  id,
  value,
  theme,
  onChange,
  min,
  max,
  readOnly
}: Props) {
  const [internalValue, setInternalValue] = useState(value);

  useEffect(() => {
    setInternalValue(value);
  }, [value]);

  const onInputChange = useCallback(
    (e) => {
      setInternalValue(e.target.value);

      // Remove any leading zero and convert to number
      let nextValue = parseInt(e.target.value || 0, 10);

      // Don't exceed max value
      if (nextValue > max) nextValue = max;
      onChange(nextValue);
    },
    [id]
  );

  const themer = autokey(themeable(theme));
  return (
    <input
      data-testid={`${id}-input`}
      readOnly={readOnly}
      {...themer('numberInput')}
      value={internalValue}
      onChange={onInputChange}
      type="number"
      min={min}
      max={max}
      step={1}
    />
  );
}

export { NumberInput };
