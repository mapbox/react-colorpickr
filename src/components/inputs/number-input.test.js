import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { NumberInput } from './number-input';
import userEvent from '@testing-library/user-event';

describe('NumberInput', () => {
  let wrapper;
  const props = {
    id: 'label',
    theme: {},
    onChange: jest.fn(),
    min: 1,
    max: 100,
    value: 10
  };

  beforeEach(() => {
    wrapper = render(<NumberInput {...props} />);
  });

  test('renders', () => {
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  test('onChange', () => {
    const mockEvent = {
      target: {
        value: '12'
      }
    };
    const input = screen.getByLabelText('label');
    userEvent.click(input);
    fireEvent.change(input, mockEvent);
    expect(props.onChange).toHaveBeenCalledWith('label', 12);
  });

  test('max caps off the value', () => {
    const mockEvent = {
      target: {
        value: '120'
      }
    };
    const input = screen.getByLabelText('label');
    userEvent.click(input);
    fireEvent.change(input, mockEvent);
    expect(props.onChange).toHaveBeenCalledWith('label', 100);
  });

  test('leading zero is chopped off', () => {
    const mockEvent = {
      target: {
        value: '012'
      }
    };
    const input = screen.getByLabelText('label');
    userEvent.click(input);
    fireEvent.change(input, mockEvent);
    expect(props.onChange).toHaveBeenCalledWith('label', 12);
  });
});
