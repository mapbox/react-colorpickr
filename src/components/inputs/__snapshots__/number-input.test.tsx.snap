// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput renders 1`] = `
<body>
  <div>
    <div>
      <label
        for="label"
      >
        label
      </label>
      <input
        id="label"
        max="100"
        min="1"
        step="1"
        type="number"
        value="10"
      />
    </div>
  </div>
</body>
`;
